Hide/add/order libs

dterrahe opened the pull request:
- reorder utility modules by drag&drop (hold CTRL+SHIFT if you want to avoid expanding/collapsing)
- restructure iop dragging as well (breakage?). Auto-scroll underlying modules and don't get confused by text entry drop targets.
- double-click on empty area in side panels to hide or show modules. This replaces options in collections/recently used collections preferences dialogs to do the same for them.
- add button in preferences/general to reset all panel options to default.
- add warning first time user pressed Tab, so they will know how to get panels back.

jenshannoschwalm commented:
Will test and review tomorrow...

victoryforce commented:
@dterrahe
Yes, indeed, if there is a tooltip that ensures discoverability, your solution is so much better! So I'm closing my PR. Haven't had a chance to do a review yet, so a few comments only after testing:

- I agree with @TurboGit's comment above that it is better to open all context menus with the right button for consistency.

- The collection and recently used collections module in darkroom works, when selecting another collection, its thumbnails appear in the filmstrip. The image opened in the darkroom does not change, of course. Therefore, we have a disconnect between this image and the content of the filmstrip. It's a little confusing at first, but you can get used to it. This is a small price for the ability to switch to another collection without jumping to the lighttable and back, which is very convenient.

- Regarding the metadata and styles modules - I didn't see any problems. At first glance, they work.

- Collection filters work in the sense that the restrictions introduced by them are reflected in the filmstrip. But if the current image does not match the filter, it looks strange: the image is open and the filmstrip is empty. I doubt whether it is worth adding this utility module to darkroom.

dterrahe commented:
> Wondering if a right-click here would be better

Absolutely. Initially the menu could be pulled up anywhere mouse clicks didn't already have another meaning (which could be handle if enough modules are expanded that there is no empty area left) so actually being _different_ from the usual right-click was an advantage to avoid conflict/confusion. But now consistency takes precedence.

Additional commit pushed to make this change. I can squash a few of the back&forth commits before merging if you want.

BTW: now that the tooltip is only shown over the empty panel area, the "fix" for unwanted tooltips over the filter range widgets isn't strictly necessary. But since it also fixes a bug (for me) where the hide tooltips toggle didn't work under WSL2 (which is a variant of (X)wayland) I'm still including it here. It also hides tooltips when pressing any modifier, so people might not need the toggle as often anymore. A while back tooltips were already suppressed when a mouse button was pressed, so dragging etc isn't obscured by them and now you can, for example, alt+scroll a slider in exposure without the tooltip covering the calculated result.

dterrahe commented:
> your solution is so much better!

thank you!

> * it is better to open all context menus with the right button for consistency

agreed. Done.

> * styles modules - I didn't see any problems

I haven't checked, but I could imagine that creating a style after just making an adjustment might not pick up the latest changes. The history is saved "frequently" but not 100% continuously. Not a problem when using styles in lighttable, because all changes will have been committed when leaving the darkroom. But within darkroom itself?

Anyway, I don't expect any unfixable problems; it's just that I (and others) haven't tested the module in this context so small fixes _might_ be needed.

> * But if the current image does not match the filter, it looks strange

This situation can already be created by changing a filter in the top bar. In fact, the selection of filters shown there can be edited in the full "collection filters" module (toggling visibility in top bar without having to re-add from a menu), so that might be one reason why someone might want to have access to the full module. Or they might prefer to have the top bar hidden but still have access to the rating filter from the left panel while doing their culling (which some people prefer to do in the darkroom. The fact that the current, or previously current, image no longer shows in the filmstrip after rejecting it is why you need access to filter toggles).

In general, this PR just intends to empower the user by lifting somewhat arbitrary restrictions. We still provide a "curated" list of modules by default, which we think make most sense in each context, but if the user wants to remove or add some; more power to them. They are trusted to deal with the consequences, rather than being told what they can and cannot have. We _are_ responsible for avoiding crashes.

And if the user _does_ get confused/messes up, there's the button in preferences/general to help them get back to a known sensible state.

victoryforce commented:
> And if the user _does_ get confused/messes up, there's the button in preferences/general to help them get back to a known sensible state.

From my experience of communicating with users (and with darktable users in particular) I can say that a certain part of them never look into the program settings. Just never looks. After all, the program works as it should? Why then look for unnecessary mental work and change some settings?

So I have a proposal for you to consider. I think it would be useful to put a "reset to default" button below the hide/show menu.

How do you feel about it?

dterrahe commented:
Easy to do.

Question is whether it should clear all settings across all views, like current preferences/general button does, or _only_ the module visibilities and position in the current view (which is what the rest of the pop-up menu is about). I'd think the latter and so I would keep both methods. 

Also, if the side panels are accidentally hidden, the menu is not accessible either so that would not be a way to get it back.

dterrahe commented:
> i can't find a way to hide a module like \"export\

Did you see that double-click changed to right-click? Do you succeed to call up the menu at all?

jenshannoschwalm commented:
Nope. Maybe I'm just blind on how-to-use? Or maybe the loupedeck interferes?

victoryforce commented:
> Question is whether it should clear all settings across all views, like current preferences/general button does, or _only_ the module visibilities and position in the current view (which is what the rest of the pop-up menu is about). I'd think the latter and so I would keep both methods.

The only logical action for such a button is to reset only the menu settings above it. And yes, it's not about replacing the button in preferences, it's about a convenient button right next to the modifiable list of modules.

dterrahe commented:
![image](https://github.com/user-attachments/assets/877fca04-197e-41e5-bf32-49631980f1d8)
There's a bug in lighttable where reloading the current view (with restored module selection) causes the center to go blank. Can be fixed by switching to culling mode but I'll need to debug what's causing this.

dterrahe commented:
> Or maybe the loupedeck interferes?

How?

So to be clear; right-clicking on an empty area in the side panel (like underneath the image information module) does _not_ open a menu for you? That would be very strange.

jenshannoschwalm commented:
> So to be clear; right-clicking on an empty area in the side panel (like underneath the image information module) does _not_ open a menu for you? That would be very strange.

Sorry for the noise, it was just a rebuild problem ...

dterrahe commented:
> There's a bug in lighttable where reloading the current view (with restored module selection) causes the center to go blank.

This gets interesting/tedious. When reloading the view to implement the reset module layout (and panel visibility/sizes etc) I don't want to call view::enter (because that causes extra work/problems in darkroom). However lighttable::enter is where the visibility of the filmstrip is switched off if it is not needed. So now it is always visible and when it draws itself it steals the thumbtable from the center view...

So maybe I have to amend `dt_view_manager_switch_by_view` to _only_ reload the side panel(s) when staying in the same view (and have that called from `dt_ui_restore_panels`). Or maybe the adding modules to the panels logic has to be completely refactored.

_Also_ interesting: `_get_lib_view_path` saves a separate setting for each "mode" in the lighttable. So in theory, you can have different panel widths when in filemanager/zoomable/culling mode. _And_ also different selections of modules or locations (which becomes a mess). And of course those aren't always correctly switched when switching mode. So if you change panel width in one mode, then switch to another and then go in and out of darkroom, you might be returned to a different panel width than you left.

Tbh I don't know if I want to dive into this, because for my use case it is irrelevant, but it will confuse "the user" if they make changes in different modes in lighttable or sometimes switch in and out of it in a different mode than where they made their module configuration. I don't even know what the "correct" solution would be; have visibility saved only lightroom wide (as opposed to other modules which depend on mode) or allow users to have different modules while culling (and force everybody else to duplicate their config across all modes). And I'm not interested in the discussion.

#13278 wants to introduce a whole new lighttable "mode" just to have different submodes when culling, so that makes everything even _more_ interesting.

dterrahe commented:
Pushed a fix for problems in lighttable when restoring default panel configuration. Now you can hide/show any collapsible module, including lua scripts installer. The lighttable has only one module config for all (culling etc) modes.

Release Notes:
- Drag&drop utility module headers to reposition them across the left and right panels. Hold ctrl+shift to avoid expanding/collapsing them before dragging. Each view can have a different layout.
- Right-click on the empty panel area below the modules to get a menu where they can be hidden or shown. This allows additional modules to be added to the darkroom, like _metadata editor_ and _styles_.
- The menu also contains an option "restore defaults" that resets the selection and position of modules in the current view. In the preferences dialog, on the general tab, there's a "reset view panels" button that resets _all_ views, including visibility and width of the panels themselves.
- The first time a new user presses Tab, they will be warned that this will hide all panels and how to get them back. Hopefully this prevents some confusion or frustration.
- Drag&drop of processing modules in the darkroom right panel has been improved to auto-scroll when reaching the top or bottom and to not get confused when images get dragged into the area.

@TurboGit I suggest you look at https://github.com/dterrahe/darktable/commit/605a725fbc6b9da7f64249963f43e2618ddb6f11 too, because it improves the direction of lines pointing at modules when they may have been moved.

dterrahe commented:
> do you think it would be possible to close it with a right-click too

This was a head scratcher. Neither right or left clicking on an empty panel area would close _any_ menu (it would just temporarily put them in the background). I had to capture both button press _and_ release events (even if not attaching a handler to the latter) to make this work again. Go figure.

In the process I fixed a bunch of memory leaks around showing and closing popup menus (they were not destroyed/released, except if they were stored in `darktable.gui->presets_popup_menu`; then they would be destroyed first before creating the next menu). The correct way of doing this is to `unref` when the "deactivate" signal is sent. Cleaned up the existing workaround.

dterrahe commented:
fixes and rebase

dterrahe commented:
> ERROR in lualib, couldn't find a position

That's because the list of all modules is being sorted whether they are visible or not. I could just create a separate list of all the visible ones in `dt_view_manager_switch_by_view` and then sort that. And not bother with keeping the list sorted anywhere else.

wpferguson commented:
@dterrahe there are 24 lua scripts that create libs.  A couple of them run in darkroom view (ext_editor and maybe another one or two).  When I use script_manager to stop a script that has a lib, I just turn off the lib visibility and mark it not to start the next time darktable is started.

Will manipulating lib visibility from lua cause problems?

dterrahe commented:
We're going to see a few bug reports.
- triggered by @wpferguson I realised that switching on the navigator/scopes/others in darkroom won't be possible unless they were on when switching to the darkroom (i.e. on->off->on will work, but not off->on)
- in darkroom dragging a lib (left panel) over an iop (right panel) causes a crash. In other views you can drag libs from left to right.

Especially the first issue will need some restructuring/rethinking/trying/testing today. Working on it.

dterrahe commented:
> ERROR in lualib, couldn't find a position

@wpferguson the simplest solution would be to just remove the warning; the code suggests it is not really a problem, just unexpected. Not unexpected anymore, so...

Because the reordering code _does_ expect the lib list to be ordered at all times (with invisible modules at either end, so 0 is fine).

dterrahe commented:
> Will manipulating lib visibility from lua cause problems?

Not "problems" per se. But since the module will only be hidden in the current view, switching to another view might show it again. And since the module is still in the list, it can be shown again via the right-click panel menu (but of course after restart it won't be available). Also, when repositioning libs with some hidden modules in between, you might get some benign strangeness. Nothing that can be sorted out after the restart.
